Adding Webpack v3 and Babel for ES6 Support

greteltrost opened the pull request:
Added gulp task `scripts` and webpack is handled with this task. If everything is ok I will update README to reflect the changes.

Steps to test:

1- install fresh drupal
2- clone repo (github.com/greteltrost/zen) and checkout to 'webpack' branch
3- install new theme from starterkit and enable it
4- test es6 by adding some sample code on js/init.js:

```js
const double = [1,2,3].map((num) => num * 2);
```

5- build js with webpack by running `gulp` or `NODE_ENV=production gulp` for minified version (or only javascript with `gulp scripts`)
6- check new code, should be transpiled to something that all browsers should be capable to run on /dist/app.js

```js
"use strict";
var double = [1, 2, 3].map(function (num) {
  return num * 2;
});
```

More info about webpack on [https://webpack.js.org/concepts/](Webpack)

Commit:
Adding Webpack v2 and Babel for ES6 Support

greteltrost commented:
Note: I added lodash as example, but commented it out on init.js

pabloguerino commented on STARTERKIT/webpack.config.js:
```diff
@@ -0,0 +1,29 @@
+  },
+  devtool: 'source-map'
+};
+
```
Also related to #5, we need to handle the `NODE_ENV` variable here when it gets the value of `production` and add uglifyjs as a webpack plugin,

something like:
```js
if (process.env.NODE_ENV === 'production') {
    config.plugins.push(
        new webpack.optimize.UglifyJsPlugin({
            compress: {
                screw_ie8: true
            }
        })
    )
}
```

greteltrost replied:
Updated webpack to use env variables and handle file compression.

andriyun commented:
Great work!
Thank you @xArturia
I'd like to test and check how it works.

I have no experience with babel. So could you provide some simple steps to check like:

* install fresh drupal
* clone repo with this PR changes
* install new theme from starter kit and enable
* add few lines of ES6 code to [filename]
* build js using gulp
* check new code

Thank you )

greteltrost commented:
Updated PR description to add steps for testing and some more info. Let me know if you run into any issues.

andriyun commented on STARTERKIT/gulpfile.js:
```diff
@@ -119,15 +120,25 @@ var gulp      = require('gulp'),
+// Build Javascript.
+// #################
+gulp.task('scripts', [], function () {
+  return gulp.src('js/init.js')
```
Here would be better use options.theme.js variable

andriyun commented on STARTERKIT/js/script.js:
```diff
@@ -1,31 +0,0 @@
-/**
```
Why we need to remove script.js file?
This file will be included in libraries definition
https://github.com/skilld-labs/zen/blob/8.x-7.x/STARTERKIT/STARTERKIT.libraries.yml#L15

pabloguerino replied:
it suits us better because init.js makes more sense as its the initialization file

andriyun commented on STARTERKIT/webpack.config.js:
```diff
@@ -0,0 +1,46 @@
+  },
+  output: {
+    path: __dirname + '/dist',
+    filename: 'app.js',
```
How we will include this file to project theme?
Maybe need to add reference to this file in libraries file  below https://github.com/skilld-labs/zen/blob/8.x-7.x/STARTERKIT/STARTERKIT.libraries.yml#L15

Btw app.js name has weird semantic.
For me better will be combine/replace this file with original scripts file

pabloguerino replied:
its the generated output, we add only `app.js` and we are done

andriyun commented:
Need final affection tests of es6 code to project theme.
It can be simple case like bind click event to any dom element etc.

greteltrost commented:
Updated `scripts` task with options.theme.js, Thanks Andriy I missed to update that.
`script.js` file has been renamed to `init.js` as it makes more sense to me, because this file will only initialize all packages and classes you may need.
For example, to add the standard behaviors file I added an example:
```js
require('script-loader!./behaviors.js');
```
This way we can have the old `theme.behaviors.js` file but parsed by babel (so, with ecmascript6 features)
And you can also create new classes, but this covers way more than what I can write in a few lines:
you can load your own packages (to separate logic between files)
```js
import * from 'sidebar.js'
```
and then webpack will parse all of them for you and you will get only one file in result (this is NOT concatenation, webpack is way more advanced than that)

andriyun commented:
We still have not included js files to theme :(

Before changes we have script.js file that can be included to theme after comment line https://github.com/skilld-labs/zen/blob/8.x-7.x/STARTERKIT/STARTERKIT.libraries.yml#L15 

Now this file is deleted

@xArturia lets add proper changes to STARTERKIT.libraries.yml

greteltrost commented:
Fixed bug in gulpfile and added the javascript output to the theme libraries file.

andriyun commented:
thank you Gretel
Last step is resolve conflicts

andypost commented on STARTERKIT/.babelrc:
```diff
@@ -0,0 +1,3 @@
+{
+  "presets": ["es2015"]
+}
```
NL

andypost commented on STARTERKIT/.gitignore:
```diff
@@ -5,3 +5,4 @@ components/asset-builds/sprites
 styleguide
 libraries
 *.map
+dist
```
NL

pabloguerino replied:
done

andypost commented on STARTERKIT/gulpfile.js:
```diff
@@ -122,22 +123,34 @@ options.gulpWatchOptions = {};
-  $           = require('gulp-load-plugins')(),
-  browserSync = require('browser-sync').create(),
-  del         = require('del'),
+var gulp        = require('gulp'),
```
This makes drupal code style checker `phpcs --standard=Drupal .` to fail

andypost reviewed (commented):
Codestyle issues, but adopt/exclude gulpfile from sniffers makes sense.
Maybe needs new issue to discus "buatifer" for js (compatiple with drupal.ord requirements)

pabloguerino commented:
What are we waiting for here ?
If you need help for testing, please don't hesitate to ask for instructions so we can get this merged and move forward.

andypost commented:
1) My review still not addressed
2) Drupal core discussion about es6 https://www.drupal.org/node/2809281#comment-12075716

pabloguerino commented:
it should be ready now, some of the issues (for the gulpfile) will be solved in a separated task

pabloguerino commented:
Updated and rebased, ready to be merged. This one is tested on several new projects so we know its safe to get it in.
